src/main/java/seedu/address/model/AddressBook.java

```diff
@@ -94,28 +112,67 @@ public void removeAssignment(Assignment key) {
         assignments.remove(key);
     }
 
+    //// lesson-level operations
+
+    /**
+     * Imports and adds lessons based on NUSMods timetable data.
+     */
+    public void importTimetable(TimetableData data) {
+        try {
+            clearLessons();
+            List<Lesson> lessons = TimetableRetriever.retrieveLessons(data);
+            for (Lesson lesson : lessons) {
+                addLesson(lesson);
+            }
+        } catch (IOException | ParseException e) {
+            // nothing happens for now.
+        }
+    }
+
+    /**
+     * Adds a lesson to the address book.
+     * The lesson must not already exist in the address book.
+     */
+    public void addLesson(Lesson lesson) {
+        lessons.add(lesson);
+    }
+
+    /**
+     * Clears all lessons in address book.
+     */
+    public void clearLessons() {
+        lessons.removeAll();
+    }
+
     //// util methods
 
     @Override
     public String toString() {
         return assignments.asUnmodifiableObservableList().size() + " assignments";
-        // TODO: refine later
+        // TODO: refine later (add lessons?)
     }
 
     @Override
     public ObservableList<Assignment> getAssignmentList() {
         return assignments.asUnmodifiableObservableList();
     }
 
+    @Override
+    public ObservableList<Lesson> getLessonList() {
+        return lessons.asUnmodifiableObservableList();
+    }
+
     @Override
     public boolean equals(Object other) {
         return other == this // short circuit if same object
                 || (other instanceof AddressBook // instanceof handles nulls
-                && assignments.equals(((AddressBook) other).assignments));
+                && assignments.equals(((AddressBook) other).assignments))
+                && lessons.equals(((AddressBook) other).lessons);
     }
 
     @Override
     public int hashCode() {
         return assignments.hashCode();
+        // TODO: unsure about lessons.hashCode()
     }
 }
```

Comment:
I think can just use the hashcode method you created in Lesson.java as given the way you implemented it seems similar to how AB3 implemented it for Person, I think lessons.hashCode() should be fine.

Reply:
I was thinking more of like how to "combine" assignments.hasCode() and lessons.hashCode() together

Reply:
I have changed it so that it does this: prime number 31 * sum of the two hash codes.